datalad/support/gitrepo.py

```diff
@@ -4076,6 +4056,102 @@ def _save_add(self, files, git_opts=None):
+            return v
+
+        def write_submodule_config(fobj, name, props):
+            fmt = '[submodule {_q_}{_name_}{_q_}]\n'
+            for p in props:
+                fmt += '\t{p} = {{{p}}}\n'.format(p=p)
+            fobj.write(
+                fmt.format(
+                    _q_='' if name.startswith('"') else '"',
+                    _name_=quote_config(name, section=True),
+                    **{k: quote_config(v) for k, v in props.items()}))
+
+        # first gather info from all datasets in read-only fashion, and then
+        # update index, .gitmodules and .git/config at once
+        info = []
+        for path in paths:
+            rpath = str(path.relative_to(self.pathobj).as_posix())
+            subm = GitRepo(path, create=False, init=False)
+            subm_commit = subm.get_hexsha()
+            if not subm_commit:
+                yield get_status_dict(
+                    action='add_submodule',
+                    ds=self,
+                    path=path,
+                    status='error',
+                    message=('cannot add subdataset %s with no commits', subm),
+                    logger=lgr)
+                continue
+            # make an attempt to configure a submodule source URL based on the
+            # discovered remote configuration
+            remote, branch = subm.get_tracking_branch()
+            url = subm.get_remote_url(remote) if remote else None
+            if url is None:
+                url = './{}'.format(rpath)
+            subm_id = subm.config.get('datalad.dataset.id', None)
+            info.append(
+                dict(path=path, rpath=rpath, commit=subm_commit, id=subm_id,
+                     url=url))
+
+        # bypass any convenience or safe-manipulator for speed reasons
+        # use case: saving many new subdatasets in a single run
+        with (self.pathobj / '.gitmodules').open('a') as gmf, \
+             (self.pathobj / '.git' / 'config').open('a') as gcf:
+            for i in info:
+                self.call_git([
+                    'update-index', '--add', '--replace', '--cacheinfo', '160000',
+                    i['commit'], i['rpath']
+                ])
+                gmprops = dict(path=i['rpath'], url=i['url'])
+                if i['id']:
+                    gmprops['datalad-id'] = i['id']
+                write_submodule_config(
+                    gmf, i['rpath'], gmprops)
+                write_submodule_config(
+                    gcf, i['rpath'], dict(active='true', url=i['url']))
+
+                # This mirrors the result structure yielded for
+                # to_stage_submodules below.
+                yield get_status_dict(
+                    action='add',
+                    refds=self.pathobj,
+                    # MIH: really? why not 'dataset', it seems "add" of an existing
+                    # subdataset (i.e. update) does that, and this one mimiks it
+                    # still, two wrongs don't make a right.
```

Comment:
How to handle this was brought up explicitly at the time.

https://github.com/datalad/datalad/pull/3398#discussion_r282928154

Ideally I think we'd switch to using `type=dataset` for adding both modified and new datasets.  The main obstacle I see is that the result values are exposed via result hooks, so we need to worry about breakage.

In my view this comment doesn't add any value, though.

Reply:
I will replace this comment with a link to this statement